File: design-patterns/go/singleton/db-connection.go
package main

import "fmt"

type connection struct {
	host     string
	port     string
	username string
	password string
}

type DbConnection interface {
	printInfo()
	executeQuery(query string)
}

var dbConnectionInstance *connection

func GetDbInstance(host, port, username, password string) *connection {
	if dbConnectionInstance == nil {
		fmt.Println("Creating a new database instance")

		dbConnectionInstance = &connection{host, port, username, password}

		return dbConnectionInstance
	}

	fmt.Println("Returing existing database instance")
	return dbConnectionInstance
}

func (conn *connection) printInfo() {
	fmt.Println("host:", conn.host)
	fmt.Println("post:", conn.port)
	fmt.Println("username:", conn.username)
	fmt.Println("password:", conn.password)
}

func (conn *connection) executeQuery(query string) {
	fmt.Println("Executing: ", query)
}

func main() {
	dbInstance1 := GetDbInstance("localhost", "3306", "root", "complex#password")
	dbInstance1.printInfo()

	dbInstance2 := GetDbInstance("localhost2", "2222", "root2", "complex#password2")
	dbInstance2.printInfo()

	if dbInstance1 == dbInstance2 {
		fmt.Println("dbInstance 1 and 2 are same instance")
	} else {
		fmt.Println("dbInstance 1 and 2 are different")
	}

}

File: design-patterns/ts/singleton/db-connection.ts
class DbConnSingleton {
    private static singletonInstance: DbConnSingleton;
    private host: string;
    private port: number;
    private username: string;
    private password: string;

    private constructor(host: string, port: number, username: string, password: string) {
        this.host = host;
        this.port = port;
        this.username = username;
        this.password = password;

    }

    public static getInstance(host: string, port: number, username: string, password: string): DbConnSingleton {
        if (this.singletonInstance) {
            console.log('Using existing db connection instance');
            return this.singletonInstance;
        }

        console.log('Creating new db connection instance');
        this.singletonInstance = new DbConnSingleton(host, port, username, password);

        return this.singletonInstance;
    }

    public printConnectionDetails() {
        console.log("host:" + this.host);
        console.log("port:" + this.port);
        console.log("username:" + this.username);
        console.log("password:" + this.password);
    }
    public executeQuery(query: string) {
        console.log("Executing query: " + query);
    }

}

const dbConnOne = DbConnSingleton.getInstance('loclahost', 1234, 'root', 'secret!pass');

console.log("DB connection details for dbConnOne:");
dbConnOne.printConnectionDetails();

const dbConnTwo = DbConnSingleton.getInstance('192.168.55.55', 2222, 'root2', 'secret!pass2');

console.log("DB connection details for dbConnTwo:");
dbConnTwo.printConnectionDetails();